### PR TITLE
[react-native] Fix Hermes version mismatch

### PR DESCRIPTION
**Pull request by gabrieldonadel**
#### Please select one of the following
- [ ] I am removing an existing difference between facebook/react-native and microsoft/react-native-macos :thumbsup:
- [ ] I am cherry-picking a change from Facebook's react-native into microsoft/react-native-macos :thumbsup:
- [x] I am making a fix / change for the macOS implementation of react-native
- [ ] I am making a change required for Microsoft usage of react-native

## Summary:

The hermes script ([hermes-utils.rb](https://github.com/microsoft/react-native-macos/compare/main...gabrieldonadel:fix-hermes-version-mismatch?expand=1#diff-ff9ad29b570f869cc7f919efbe9d306bd5f90e8fa17dd6b51f8eb83d15c120e4)) uses the react-native-macos version to check for available hermes artifacts on maven but the problem is that whenever react-native-macos version gets ahead of react-native there are no prebuilds to be found and hermes will try to build from source (which does not work). 

This PR fixes this by checking the latest available patch version for a given version and using that as the hermes-version 

## Changelog:

[MACOS] [FIXED] - Fix Hermes version mismatch

## Test Plan:

Enable hermes on rntester, set react-native package.json version to 0.73.2, run pod install and certify it's downloading the prebuild tarball

### CODE REVIEW

File: packages/react-native/sdks/hermes-engine/hermes-utils.rb [hunk 1]
```diff
@@ -203,7 +205,7 @@ def hermestag_file(react_native_path)
 def release_tarball_url(version, build_type)
     # Sample url from Maven:
     # https://repo1.maven.org/maven2/com/facebook/react/react-native-artifacts/0.71.0/react-native-artifacts-0.71.0-hermes-ios-debug.tar.gz
-    return "https://repo1.maven.org/maven2/com/facebook/react/react-native-artifacts/#{version}/react-native-artifacts-#{version}-hermes-ios-#{build_type.to_s}.tar.gz"
+    return "#{REACT_NATIVE_ARTIFACTS_MAVEN_URL}/#{version}/react-native-artifacts-#{version}-hermes-ios-#{build_type.to_s}.tar.gz"
```
Comment (Saadnajmi) [inline packages/react-native/sdks/hermes-engine/hermes-utils.rb:208]: Is this change necessary? If not, I'd prefer to make it upstream to keep diffs lower

File: packages/react-native/sdks/hermes-engine/hermes-engine.podspec [hunk 2]
```diff
@@ -10,7 +10,8 @@ react_native_path = File.join(__dir__, "..", "..")
 
 # package.json
 package = JSON.parse(File.read(File.join(react_native_path, "package.json")))
-version = package['version']
+rn_version = package['version']
+version = findLastestVersionWithArtifact(rn_version) || rn_version
```
Comment (Saadnajmi) [inline packages/react-native/sdks/hermes-engine/hermes-engine.podspec:14]: Could we mark this diff with " # [macOS / # macOS] " tags as documented here? docs/DiffsWithUpstream.md

File: packages/react-native/sdks/hermes-engine/hermes-utils.rb [hunk 3]
```diff
@@ -233,6 +235,27 @@ def resolve_url_redirects(url)
     return (`curl -Ls -o /dev/null -w %{url_effective} \"#{url}\"`)
 end
 
+# react-native-macos does not publish macos specific hermes artifacts
+# so we attempt to find the latest patch version of the iOS artifacts and use that
+def findLastestVersionWithArtifact(version)
```
Comment (Saadnajmi) [inline packages/react-native/sdks/hermes-engine/hermes-utils.rb:240]: Same comment about adding diff tags to method.

**Review by Saadnajmi — CHANGES_REQUESTED** [review]

File: packages/react-native/sdks/hermes-engine/hermes-utils.rb (hunk 1 above)
Comment (gabrieldonadel) [reply packages/react-native/sdks/hermes-engine/hermes-utils.rb]: reverted

File: packages/react-native/sdks/hermes-engine/hermes-engine.podspec (hunk 2 above)
Comment (gabrieldonadel) [reply packages/react-native/sdks/hermes-engine/hermes-engine.podspec]: Sure, updated

File: packages/react-native/sdks/hermes-engine/hermes-utils.rb (hunk 3 above)
Comment (gabrieldonadel) [reply packages/react-native/sdks/hermes-engine/hermes-utils.rb]: fixed

File: packages/react-native/sdks/hermes-engine/hermes-utils.rb [hunk 4]
```diff
@@ -5,6 +5,7 @@
 
 require 'net/http'
 require 'rexml/document'
+require 'open3'
```
Comment (Saadnajmi) [inline packages/react-native/sdks/hermes-engine/hermes-utils.rb:8]: Can you also add a diff tag to the end of this line?
```suggestion
require 'open3' # [macOS]
```

**Review by Saadnajmi — APPROVED** [review]

File: packages/react-native/sdks/hermes-engine/hermes-utils.rb (hunk 4 above)
Comment (gabrieldonadel) [reply packages/react-native/sdks/hermes-engine/hermes-utils.rb]: Opps, fixed

**Comment by gabrieldonadel** [comment]
Hi @Saadnajmi, Is there something else that needs to be addressed here to get this merged?

**Comment by Saadnajmi** [comment]
> Hi @Saadnajmi, Is there something else that needs to be addressed here to get this merged?

Nope, just a nudge! If you also want this for `0.73`, could you cherry-pick the change and make a PR against the `0.73` branch? It's easier for me to approve/merge that way
